File: Level2/다리를 지나는 트럭.py
bridge_length = 2
weight = 10
truck_weights = [7, 4, 5, 6]


def solution(bridge_length, weight, truck_weights):
    bridge = [0] * bridge_length
    count = 0

    while bridge:
        count += 1
        bridge.pop(0)
        if truck_weights:
            if sum(bridge) + truck_weights[0] <= weight:
                bridge.append(truck_weights.pop(0))
            else:
                bridge.append(0)
    return count

did = solution(bridge_length, weight, truck_weights)
print(did)

File: Level2/다리를 지나는 트럭.java
import java.util.*;
class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int time = 0;
        int idx = 0;
        int now_weight = 0;
        Queue<Integer> done = new LinkedList<>();
        Queue<Integer> on = new LinkedList<>();
        
        while(idx < truck_weights.length) {
            if(on.size() < bridge_length) {
                if(now_weight + truck_weights[idx] <= weight) {
                    on.add(truck_weights[idx]);
                    now_weight += truck_weights[idx++];
                } else 
                    on.add(0);
                time++;
            } else {
                done.add(on.peek());
                now_weight -= on.poll();
            }
        }
        if(!on.isEmpty()) 
            time += bridge_length;
        return time;
    }
}
